src/databricks/labs/ucx/azure/locations.py

```diff
@@ -0,0 +1,170 @@
+        return prefix_mapping_write, prefix_mapping_read
+
+    def _create_location_name(self, location_url: str) -> str:
+        # generate the UC external location name
+        before_at, _, after_at = location_url.partition('@')
+        container_name = before_at.removeprefix("abfss://")
+        res_name = after_at.replace(".dfs.core.windows.net", "").rstrip("/").replace("/", "_")
+        return f"{container_name}_{res_name}"
+
+    def _create_external_location(
+        self, location_url: str, prefix_mapping_write: dict[str, str], prefix_mapping_read: dict[str, str]
+    ) -> str | None:
+        location_name = self._create_location_name(location_url)
+
+        # get container url as the prefix
+        parsed_url = urlparse(location_url)
+        container_url = f"{parsed_url.scheme}://{parsed_url.netloc}/"
+
+        # try to create external location with write privilege first
+        if container_url in prefix_mapping_write:
+            self._ws.external_locations.create(
+                location_name, location_url, prefix_mapping_write[container_url], comment="Created by UCX"
+            )
+            return location_url
+        # if no matched write privilege credential, try to create read-only external location
+        if container_url in prefix_mapping_read:
+            try:
+                self._ws.external_locations.create(
+                    location_name,
+                    location_url,
+                    prefix_mapping_read[container_url],
+                    comment="Created by UCX",
+                    read_only=True,
+                )
+            except PermissionDenied as denied:
+                if "No file available under the location to read" in str(denied):
+                    # Empty location will cause failed READ permission check with read-only credential
+                    # Skip skip_validation in this case
+                    self._ws.external_locations.create(
+                        location_name,
+                        location_url,
+                        prefix_mapping_read[container_url],
+                        comment="Created by UCX",
+                        read_only=True,
+                        skip_validation=True,
+                    )
+                    return location_url
+                raise denied
+            return location_url
+        # if no credential found
+        return None
+
+    def run(self):
+        # list missing external locations in UC
+        missing_locs = [loc.location for loc in self._hms_locations.match_table_external_locations()[1]]
+
+        # get prefix to storage credential name mapping
+        prefix_mapping_write, prefix_mapping_read = self._prefix_credential_name_mapping()
+
+        # if missing external location is in prefix to storage credential name mapping
+        # create a UC external location with mapped storage credential name
```

Comment:
Missing handling for external locations that are already there - what if this command runs twice? What if locations are created by terraform or by hand? What if only part of them are created by hand?

Reply:
The `match_table_external_locations()` will first call UC apt to list existed UC external locations. So the returned `missing_locs` will not include any created UC external locations no matter how many times this command runs.